community/ndp/messaging-v1/src/main/java/org/neo4j/ndp/messaging/v1/infrastructure/ValuePath.java

```diff
@@ -19,127 +19,441 @@
+ * takes the form of an integer, which in the case of nodes simply references
+ * the zero-based index of that node.
+ * <p/>
+ * Relationship pointers are less straightforward, using a one-based indexing
+ * system (i.e. relationship #1 refers to the item 0 from the relationship
+ * list. Additionally, the sign of the index is relevant: a positive index
+ * denotes that the relationship was traversed with its direction, a
+ * negative index denotes that the relationship was traversed against its
+ * direction. Note that relationship items do not contain endpoint information,
+ * this is described by this sequential context.
+ * <p/>
+ * Finally, the first sequence value will always be 0 by definition. Therefore,
+ * this value is never actually transmitted and is instead reconstituted by the
+ * receiving agent.
+ * <p/>
+ * Take as an example, the following path:
+ * <pre>
+ *     (a)--&gt;(b)&lt;--(a)--&gt;(c)--&gt;(d)
+ * </pre>
+ * Here, the nodes would be uniquely listed as [(a), (b), (c), (d)] and the
+ * relationships as [(a:b), (a:c), (c:d)]. Applying zero-based and one-based
+ * indexing to these lists respectively, we get:
+ * <pre>
+ *       0    1    2    3         1      2      3
+ *     [(a), (b), (c), (d)]    [(a:b), (a:c), (c:d)]
+ * </pre>
+ * Applying these indexes to the original path then gives us:
+ * <pre>
+ *               (a)--&gt;(b)&lt;--(a)--&gt;(c)--&gt;(d)
+ *     Node ID :  0     1     0     2     3
+ *      Rel ID :     1     1     2     3
+ * </pre>
+ * And negating relationships that are traversed backwards, we get:
+ * <pre>
+ *               (a)--&gt;(b)&lt;--(a)--&gt;(c)--&gt;(d)
+ *     Node ID :  0     1     0     2     3
+ *      Rel ID :    +1    -1    +2    +3
+ * </pre>
+ * Finally, we can drop the leading 0 from the Path structure and combine the
+ * sequence information with the other path data. This results in an overall
+ * Path representation as below:
+ * <pre>
+ *     Structure Path 'P' {
+ *         nodes         = [(a), (b), (c), (d)]
+ *         relationships = [(a:b), (a:c), (c:d)]
+ *         sequence      = [+1, 1, -1, 0, +2, 2, +3, 3]
+ *     }
+ * </pre>
+ * Therefore, for the original path containing five nodes and four relationships,
+ * we end up transmitting only four nodes, three unbound relationships and eight
+ * bytes of sequence data (small integers require only a single byte in PackStream).
+ */
 public class ValuePath implements Path
 {
-    private final PropertyContainer[] entities;
+    private static final int STRUCT_FIELD_COUNT = 5;
+
+    private static List<Node> packNodes( Neo4jPack.Packer packer, Path path )
+            throws IOException
+    {
+        // TODO: rewrite this without object creation
+        // TODO: (might require additions to the Path interface)
+        List<Node> nodes = new ArrayList<>( path.length() + 1 );
+        for ( Node node : path.nodes() )
+        {
+            if ( !nodes.contains( node ) )
```

Comment:
This is an O(n) operation with ArrayList - we should use `HashSet` here, otherwise this will be very slow for longer paths.